stackrox/stackrox
# ROX-25739: Sentinel notifier support client cert authentication

### Description

This PR introduces client certificate authentication for Microsoft Sentinel.

To use client cert authentication a user must:

- configure a client  cert
- configure a private key

If a user configures a `secret` and `clientCertAuthConfig` the client certificate is preferred. One of secret or clientCertAuthConfig must be configured.

```
// client certificate which is used for authentication
message ClientCertAuthConfig {
  // PEM encoded ASN.1 DER format.
  string client_cert = 1;

  // PEM encoded PKCS #8, ASN.1 DER format.
  string private_key = 2;
}
ClientCertAuthConfig client_cert_auth_config = 7;
```

### User-facing documentation


### Testing and quality

 - CI and unit tests
 - Used a testing script to sent notifications to Sentinel with client cert authentication enabled:

 1. Generate key and certificate `openssl req -x509 -newkey rsa:2048 -days 365 -keyout ca-key.pem -out ca-cert.pem -nodes`
 2. Export variables as in bash script
 3. Run script, you should see `{}` as a response and HTTP `200` status code

```
#!/usr/bin/env bash

logIngestionEndpoint=$SENTINEL_LOG_INGESTION_ENDPOINT
dataCollectionRuleId=$DATA_COLLECTION_RULE_ID
streamName=$STREAM_NAME
directoryTenantID=$DIRECTORY_TENANT_ID
applicationClientID=$APP_CLIENT_ID
secret=$SECRET
clientCert=$CERT
key=$KEY


notifier=$(cat <<EOF
{
   "name": "sentinel-notifier",
   "uiEndpoint": "localhost:8000",
   "type": "microsoftSentinel",
   "microsoftSentinel": {
       "logIngestionEndpoint": "$logIngestionEndpoint",
       "alertDcrConfig": {
           "dataCollectionRuleId": "$dataCollectionRuleId",
           "streamName": "$streamName",
           "enabled": true
       },
       "auditLogDcrConfig": {
           "dataCollectionRuleId": "$dataCollectionRuleId",
           "streamName": "$streamName",
           "enabled": true
       },
       "directoryTenantId": "$directoryTenantID",
       "applicationClientId": "$applicationClientID",
       "clientCertAuthConfig": {
         "clientCert": "$clientCert",
         "privateKey": "$key",
       },
   }
}
EOF
)


roxcurl /v1/notifiers/test -X POST --data $notifier -H "Content-Type: application/json
```
